File: algo_library/cpp/maxflow.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i, n) for (int i = 0; i < (int)(n); i++)
#define print_vec(v) { rep(i, v.size()) cout << v[i] << " "; cout << endl; }
#define ll long long
const int MOD = 998244353;
const int INF = 1000000000;
const ll INFll = 9223372036854775807;

struct maxflow_edge{
    int to, cap, rev_idx;
};

class maximumflow{
    public:
        vector<vector<maxflow_edge>> G;
        vector<bool> vis;

        void init(int N){
            G.resize(N+1);
            vis.resize(N+1,false);
        }

        void add_edge(int a, int b, int c){
            int ga_size = G[a].size();
            int gb_size = G[b].size();
            G[a].emplace_back(maxflow_edge{b,c,gb_size});
            G[b].emplace_back(maxflow_edge{a,0,ga_size});
        }

        int dfs(int v, int goal, int f){
            if (v==goal) return f;
            vis[v] = true;
            for (auto &edge : G[v]){
                if (edge.cap == 0) continue;
                if (vis[edge.to]) continue;
                int flow = dfs(edge.to, goal, min(f, edge.cap));
                if (flow > 0){
                    edge.cap -= flow;
                    G[edge.to][edge.rev_idx].cap += flow;
                    return flow;
                }
            }
            return 0;
        }

        int maxflow(int N, int M, int start, int goal){
            int total_flow = 0;
            while (true){
                for (int i=0; i<=N; i++) vis[i] = false;
                const int INF = 1000000000;
                int res = dfs(start,goal,INF);
                if (res==0) break;
                total_flow += res;
            }
            return total_flow;
        }
};

File: algo_library/python/graph/maxflow.py
import pypyjit
pypyjit.set_param('max_unroll_recursion=-1')
sys.setrecursionlimit(10**7)

class maxflow_edge:
  def __init__(self, to, cap, rev_idx):
    self.to = to # 行き先
    self.cap = cap # 容量
    self.rev_idx = rev_idx # 逆向きの辺のインデックス

def dfs(v,goal,f,G,vis):
  if v==goal: return f
  vis[v] = True
  for edge in G[v]:
    if edge.cap > 0 and vis[edge.to] == False:
      flow = dfs(edge.to, goal, min(f, edge.cap), G, vis)
      # フローを流せるなら、残余グラフの容量を flow 増減させる
      if flow >= 1:
        edge.cap -= flow
        G[edge.to][edge.rev_idx].cap += flow
        return flow
  # 流せないなら
  return 0

def maxflow(N,M,start,goal):
  G = [[] for _ in range(N+1)]
  for _ in range(M):
    a,b,c = ms()
    G[a].append(maxflow_edge(b,c,len(G[b])))
    G[b].append(maxflow_edge(a,0,len(G[a])-1)) # 逆向き
  total_flow = 0
  inf = pow(10,18)
  while True:
    vis = [False]*(N+1)
    res = dfs(start, goal, inf, G, vis)
    if res > 0: total_flow += res
    else: break
  return total_flow